OK, further tests added. Is this what you are thinking of?

Aside, using the test data one finds another bug... Before I raise an issue, since you are more familiar with what has already been seen, is it already noticed somewhere?
```
descr = [('v', int, 3), ('x', [('y', float)])]
x = array([
         [([1, 2, 3], (1.0,)),  ([1, 2, 3], (2.0,))],
         [([1, 2, 3], (3.0,)),  ([1, 2, 3], (4.0,))]], dtype=descr)
x == x
ValueError                                Traceback (most recent call last)
<ipython-input-15-1238b007ad2c> in <module>()
----> 1 x == x

/data/mhvk/venv/numpy-dev/lib/python3.6/site-packages/numpy-1.15.0.dev0+924ad94-py3.6-linux-x86_64.egg/numpy/ma/core.py in __eq__(self, other)
   4013         and other considered equal only if both were fully masked.
   4014         """
-> 4015         return self._comparison(other, operator.eq)
   4016 
   4017     def __ne__(self, other):

/data/mhvk/venv/numpy-dev/lib/python3.6/site-packages/numpy-1.15.0.dev0+924ad94-py3.6-linux-x86_64.egg/numpy/ma/core.py in _comparison(self, other, compare)
   3973             sbroadcast = np.broadcast_to(self, broadcast_shape, subok=True)
   3974             sbroadcast._mask = mask
-> 3975             sdata = sbroadcast.filled(odata)
   3976             # Now take care of the mask; the merged mask should have an item
   3977             # masked if all fields were masked (in one and/or other).

/data/mhvk/venv/numpy-dev/lib/python3.6/site-packages/numpy-1.15.0.dev0+924ad94-py3.6-linux-x86_64.egg/numpy/ma/core.py in filled(self, fill_value)
   3696         if m.dtype.names:
   3697             result = self._data.copy('K')
-> 3698             _recursive_filled(result, self._mask, fill_value)
   3699         elif not m.any():
   3700             return self._data

/data/mhvk/venv/numpy-dev/lib/python3.6/site-packages/numpy-1.15.0.dev0+924ad94-py3.6-linux-x86_64.egg/numpy/ma/core.py in _recursive_filled(a, mask, fill_value)
   2484             _recursive_filled(current, mask[name], fill_value[name])
   2485         else:
-> 2486             np.copyto(current, fill_value[name], where=mask[name])
   2487 
   2488 

ValueError: could not broadcast input array from shape (2,2) into shape (2,2,3)
```

--- a/doc/release/1.15.0-notes.rst
+++ b/doc/release/1.15.0-notes.rst
@@ -44,7 +44,11 @@ New Features
 Improvements
 ============
 
+``MaskedArray.astype`` now is identical to ``ndarray.astype``
+-------------------------------------------------------------
+
+This means it takes all the same arguments, making more code written for
+ndarray work for masked array too.
 
 Changes
 =======
-

--- a/doc/release/1.15.0-notes.rst
+++ b/doc/release/1.15.0-notes.rst
@@ -46,9 +46,9 @@ Improvements
 
 ``MaskedArray.astype`` now is identical to ``ndarray.astype``
 -------------------------------------------------------------
-
 This means it takes all the same arguments, making more code written for
 ndarray work for masked array too.
+
 
 Changes
 =======

--- a/numpy/ma/core.py
+++ b/numpy/ma/core.py
@@ -1624,7 +1624,7 @@ def make_mask(m, copy=False, shrink=True, dtype=MaskType):
 
     # Make sure the input dtype is valid.
     dtype = make_mask_descr(dtype)
-    
+
     # legacy boolean special case: "existence of fields implies true"
     if isinstance(m, ndarray) and m.dtype.fields and dtype == np.bool_:
         return np.ones(m.shape, dtype=dtype)
@@ -2978,11 +2978,30 @@ class MaskedArray(ndarray):
             # heuristic it's not bad.) In all other cases, we make a copy of
             # the mask, so that future modifications to 'self' do not end up
             # side-effecting 'obj' as well.
-            if (obj.__array_interface__["data"][0]
+            if (_mask is not nomask and obj.__array_interface__["data"][0]
                     != self.__array_interface__["data"][0]):
-                _mask = _mask.copy()
+                # We should make a copy. But we could get here via astype,
+                # in which case the mask might need a new dtype as well
+                # (e.g., changing to or from a structured dtype), and the
+                # order could have changed. So, change the mask type if
+                # needed and use astype instead of copy.
+                if self.dtype == obj.dtype:
+                    _mask_dtype = _mask.dtype
+                else:
+                    _mask_dtype = make_mask_descr(self.dtype)
+
+                if self.flags.c_contiguous:
+                    order = "C"
+                elif self.flags.f_contiguous:
+                    order = "F"
+                else:
+                    order = "K"
+
+                _mask = _mask.astype(_mask_dtype, order)
+
         else:
             _mask = nomask
+
         self._mask = _mask
         # Finalize the mask
         if self._mask is not nomask:
@@ -3139,45 +3158,6 @@ class MaskedArray(ndarray):
                 output.fill_value = fill_value
         return output
     view.__doc__ = ndarray.view.__doc__
-
-    def astype(self, newtype):
-        """
-        Returns a copy of the MaskedArray cast to given newtype.
-
-        Returns
-        -------
-        output : MaskedArray
-            A copy of self cast to input newtype.
-            The returned record shape matches self.shape.
-
-        Examples
-        --------
-        >>> x = np.ma.array([[1,2,3.1],[4,5,6],[7,8,9]], mask=[0] + [1,0]*4)
-        >>> print(x)
-        [[1.0 -- 3.1]
-         [-- 5.0 --]
-         [7.0 -- 9.0]]
-        >>> print(x.astype(int32))
-        [[1 -- 3]
-         [-- 5 --]
-         [7 -- 9]]
-
-        """
-        newtype = np.dtype(newtype)
-        newmasktype = make_mask_descr(newtype)
-
-        output = self._data.astype(newtype).view(type(self))
-        output._update_from(self)
-
-        if self._mask is nomask:
-            output._mask = nomask
-        else:
-            output._mask = self._mask.astype(newmasktype)
-
-        # Don't check _fill_value if it's None, that'll speed things up
-        if self._fill_value is not None:
-            output._fill_value = _check_fill_value(self._fill_value, newtype)
-        return output
 
     def __getitem__(self, indx):
         """
@@ -4303,7 +4283,7 @@ class MaskedArray(ndarray):
         elif self._mask:
             raise MaskError('Cannot convert masked element to a Python int.')
         return int(self.item())
-    
+
     def __long__(self):
         """
         Convert to long.
@@ -4314,7 +4294,7 @@ class MaskedArray(ndarray):
         elif self._mask:
             raise MaskError('Cannot convert masked element to a Python long.')
         return long(self.item())
-      
+
 
     def get_imag(self):
         """

--- a/numpy/ma/tests/test_core.py
+++ b/numpy/ma/tests/test_core.py
@@ -5037,10 +5037,37 @@ def test_ufunc_with_output():
     y = np.add(x, 1., out=x)
     assert_(y is x)
 
+
 def test_astype():
     descr = [('v', int, 3), ('x', [('y', float)])]
-    x = array(([1, 2, 3], (1.0,)), dtype=descr)
-    assert_equal(x, x.astype(descr))
+    x = array([
+        [([1, 2, 3], (1.0,)),  ([1, 2, 3], (2.0,))],
+        [([1, 2, 3], (3.0,)),  ([1, 2, 3], (4.0,))]], dtype=descr)
+    x[0]['v'][0] = np.ma.masked
+
+    x_a = x.astype(descr)
+    assert x_a.dtype.names == np.dtype(descr).names
+    assert x_a.mask.dtype.names == np.dtype(descr).names
+    assert_equal(x, x_a)
+
+    assert_(x is x.astype(x.dtype, copy=False))
+    assert_equal(type(x.astype(x.dtype, subok=False)), np.ndarray)
+
+    x_f = x.astype(x.dtype, order='F')
+    assert_(x_f.flags.f_contiguous)
+    assert_(x_f.mask.flags.f_contiguous)
+
+    # Also test the same indirectly, via np.array
+    x_a2 = np.array(x, dtype=descr, subok=True)
+    assert x_a2.dtype.names == np.dtype(descr).names
+    assert x_a2.mask.dtype.names == np.dtype(descr).names
+    assert_equal(x, x_a2)
+
+    assert_(x is np.array(x, dtype=descr, copy=False, subok=True))
+
+    x_f2 = np.array(x, dtype=x.dtype, order='F', subok=True)
+    assert_(x_f2.flags.f_contiguous)
+    assert_(x_f2.mask.flags.f_contiguous)
 
 
 ###############################################################################

